### PR TITLE
[REFACTOR] move blocks in file_system to /blocks dir under each file

### PR DESCRIPTION
**Pull request by jake-nyquist**
**Motivation** 
Several teams, including security and file mutability, have requested the ability to create additional znode children under each file. Currently, the system finds the block for each file by getting all children under the file path. By simply adding additional children, we would return these new child types as blocks.

**Solution**
This refactor adds a `/blocks` subdir under each file, which is where we will store the blocks. 

**Risks**
Because the designers of RDFS expected the blocks to be a child of the file, and also stored data in the file, it is important that we still store file metadata under the file, but redirect all calls looking for the blocks elsewhere. 

Because this could mess up a lot of stuff that may have variable test coverage, it would be good if people could look at it.

### CODE REVIEW

**Comment by cannontwo** [comment]
I'm confused about what this refactoring accomplishes. Could you elaborate on the need for this change?

**Comment by jake-nyquist** [comment]
@cannon10100 
Current tree:
```
├── file_system
│      └── dir_name
│               └── file_name
│                          └── block_id
```


Tree after refactor:
```
├── file_system
│   └── dir_name
│       └── file_name
│            ├── blocks
│            │   └── block_id
│            ├── _permissions_
│            │   └── _access_control_list_
│            └── _leases_
│			└── _lease_

```
The items in italics are not implemented here. This allows us to store the block_ids as children of znodes without colliding with the new children for ACLs and Leases that security and file mutability say that they need.

**Review by cannontwo — APPROVED** [review]
Diffs look fine

File: zookeeper/include/zk_nn_client.h [hunk 1]
```diff
@@ -255,9 +255,17 @@ class ZkNnClient : public ZkClientCommon {
...
+   * where the data is located
    */
-  std::string ZookeeperPath(const std::string &hadoopPath);
+  std::string ZookeeperBlocksPath(const std::string &hadoopPath);
+
+   /**
+   * Given the filesystem path, get the full zookeeper path for the dir
+   * where the file metadata is written
+   */
+  std::string ZkNnClient::ZookeeperFilePath(const std::string &hadoopPath);
```
Comment (ghost) [inline zookeeper/include/zk_nn_client.h]: i think you might not need the `ZkNnClient::` in front of this new function in this file.

**Review by ghost — APPROVED** [review]
LGTM

**Review by Antzen — APPROVED** [review]
For the reasons you stated, I agree that separating the block_id's into its own sub-directory is a good idea. 

Besides the build errors, changes overall appear fine to me. It looks like the error may involve the way you declare ZookeeperFilePath in zk_nn_client.h, as Eddie pointed out in his review.

File: zookeeper/source/zk_nn_client.cc [hunk 2]
```diff
@@ -1156,6 +1156,12 @@ std::string ZkNnClient::ZookeeperPath(const std::string &hadoopPath) {
   return zkpath;
 }
 
+std::string ZkNnClient::ZookeeperBlocksPath(const std::string &hadoopPath) {
+  std::string zkpath = ZookeeperFilePath(&hadoopPath);
```
Comment (LINKIWI) [inline zookeeper/source/zk_nn_client.cc]: Should this be

```cpp
std::string zkpath = ZookeeperFilePath(hadoopPath);
```

?

**Review by LINKIWI — APPROVED** [review]

File: zookeeper/source/zk_nn_client.cc (hunk 2 above)
Comment (jake-nyquist) [reply zookeeper/source/zk_nn_client.cc]: yes

File: zookeeper/source/zk_nn_client.cc [hunk 3]
```diff
@@ -1286,13 +1292,13 @@ bool ZkNnClient::add_block(const std::string &file_path,
...
 
-  LOG(INFO) << "Generating block for " << ZookeeperPath(file_path);
+  LOG(INFO) << "Generating block for " << ZookeeperBlocksPath(file_path);
 
   // ZooKeeper multi-op to add block
   // (Note the actual path of the znode the first create op makes will have a 9
   // digit number appended onto the end because of the ZOO_SEQUENCE flag)
   auto seq_file_block_op = zk->build_create_op(
-      ZookeeperPath(file_path + "/block_"),
+      ZookeeperBlocksPath(file_path + "/block_"),
```
Comment (riceluxs1t) [inline zookeeper/source/zk_nn_client.cc]: Shouldn't this be ZookeeperBlocksPath(file_path)

**Comment by jake-nyquist** [comment]
@eddiedugan @LINKIWI  this seems to have caused some sort of regression on test time, is that normal or should I be worried?

**Pull request merged**